File: Backend/controllers/forumController.js
const Post = require('../models/Post');
const Comment = require('../models/Comment');
const User = require('../models/User');

// @desc    Get all posts
// @route   GET /api/forum
// @access  Public
const getPosts = async (req, res) => {
  try {
    const posts = await Post.find()
      .populate('author', 'username')
      .populate({
        path: 'comments',
        populate: {
          path: 'author',
          select: 'username'
        }
      })
      .sort({ createdAt: -1 });

    res.json(posts);
  } catch (error) {
    console.error('Get posts error:', error);
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

// @desc    Get post by ID
// @route   GET /api/forum/:id
// @access  Public
const getPostById = async (req, res) => {
  try {
    const post = await Post.findById(req.params.id)
      .populate('author', 'username')
      .populate({
        path: 'comments',
        populate: {
          path: 'author',
          select: 'username'
        }
      });

    if (!post) {
      return res.status(404).json({ message: 'Post not found' });
    }

    res.json(post);
  } catch (error) {
    console.error('Get post error:', error);
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

// @desc    Create a new post
// @route   POST /api/forum
// @access  Private
const createPost = async (req, res) => {
  try {
    const { title, content } = req.body;

    const post = await Post.create({
      title,
      content,
      author: req.user._id
    });

    const populatedPost = await Post.findById(post._id)
      .populate('author', 'username');

    res.status(201).json(populatedPost);
  } catch (error) {
    console.error('Create post error:', error);
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

// @desc    Update a post
// @route   PUT /api/forum/:id
// @access  Private
const updatePost = async (req, res) => {
  try {
    const { title, content } = req.body;
    const post = await Post.findById(req.params.id);

    if (!post) {
      return res.status(404).json({ message: 'Post not found' });
    }

    // Check if user is the author of the post
    if (post.author.toString() !== req.user._id.toString()) {
      return res.status(403).json({ message: 'Not authorized to update this post' });
    }

    post.title = title || post.title;
    post.content = content || post.content;

    const updatedPost = await post.save();
    const populatedPost = await Post.findById(updatedPost._id)
      .populate('author', 'username')
      .populate({
        path: 'comments',
        populate: {
          path: 'author',
          select: 'username'
        }
      });

    res.json(populatedPost);
  } catch (error) {
    console.error('Update post error:', error);
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

// @desc    Delete a post
// @route   DELETE /api/forum/:id
// @access  Private
const deletePost = async (req, res) => {
  try {
    const post = await Post.findById(req.params.id);

    if (!post) {
      return res.status(404).json({ message: 'Post not found' });
    }

    // Check if user is the author of the post
    if (post.author.toString() !== req.user._id.toString()) {
      return res.status(403).json({ message: 'Not authorized to delete this post' });
    }

    // Delete all comments associated with the post
    await Comment.deleteMany({ post: post._id });

    // Delete the post
    await Post.deleteOne({ _id: post._id });

    res.json({ message: 'Post deleted successfully' });
  } catch (error) {
    console.error('Delete post error:', error);
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

// @desc    Add a comment to a post
// @route   POST /api/forum/:id/comments
// @access  Private
const addComment = async (req, res) => {
  try {
    const { content } = req.body;
    const post = await Post.findById(req.params.id);

    if (!post) {
      return res.status(404).json({ message: 'Post not found' });
    }

    const comment = await Comment.create({
      content,
      author: req.user._id,
      post: post._id
    });

    // Add comment to post's comments array
    post.comments.push(comment._id);
    await post.save();

    const populatedComment = await Comment.findById(comment._id)
      .populate('author', 'username');

    res.status(201).json(populatedComment);
  } catch (error) {
    console.error('Add comment error:', error);
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

// @desc    Delete a comment
// @route   DELETE /api/forum/:id/comments/:commentId
// @access  Private
const deleteComment = async (req, res) => {
  try {
    const { commentId } = req.params;
    const comment = await Comment.findById(commentId);

    if (!comment) {
      return res.status(404).json({ message: 'Comment not found' });
    }

    // Check if user is the author of the comment
    if (comment.author.toString() !== req.user._id.toString()) {
      return res.status(403).json({ message: 'Not authorized to delete this comment' });
    }

    // Remove comment from post's comments array
    const post = await Post.findById(req.params.id);
    if (post) {
      post.comments = post.comments.filter(
        comment => comment.toString() !== commentId
      );
      await post.save();
    }

    // Delete the comment
    await Comment.deleteOne({ _id: comment._id });

    res.json({ message: 'Comment deleted successfully' });
  } catch (error) {
    console.error('Delete comment error:', error);
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

// @desc    Like or unlike a post
// @route   PUT /api/forum/:id/like
// @access  Private
const likePost = async (req, res) => {
  try {
    const post = await Post.findById(req.params.id);

    if (!post) {
      return res.status(404).json({ message: 'Post not found' });
    }

    // Check if the post has already been liked by this user
    const alreadyLiked = post.likes.some(
      like => like.toString() === req.user._id.toString()
    );

    if (alreadyLiked) {
      // Unlike the post
      post.likes = post.likes.filter(
        like => like.toString() !== req.user._id.toString()
      );
    } else {
      // Like the post
      post.likes.push(req.user._id);
    }

    await post.save();

    res.json({ likes: post.likes, likeCount: post.likes.length });
  } catch (error) {
    console.error('Like post error:', error);
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

// @desc    Like or unlike a comment
// @route   PUT /api/forum/:id/comments/:commentId/like
// @access  Private
const likeComment = async (req, res) => {
  try {
    const { commentId } = req.params;
    const comment = await Comment.findById(commentId);

    if (!comment) {
      return res.status(404).json({ message: 'Comment not found' });
    }

    // Check if the comment has already been liked by this user
    const alreadyLiked = comment.likes.some(
      like => like.toString() === req.user._id.toString()
    );

    if (alreadyLiked) {
      // Unlike the comment
      comment.likes = comment.likes.filter(
        like => like.toString() !== req.user._id.toString()
      );
    } else {
      // Like the comment
      comment.likes.push(req.user._id);
    }

    await comment.save();

    res.json({ likes: comment.likes, likeCount: comment.likes.length });
  } catch (error) {
    console.error('Like comment error:', error);
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

module.exports = {
  getPosts,
  getPostById,
  createPost,
  updatePost,
  deletePost,
  addComment,
  deleteComment,
  likePost,
  likeComment
};

File: BackendJava/src/main/java/com/aiplantanalyze/controller/ForumController.java
package com.aiplantanalyze.controller;

import com.aiplantanalyze.model.Comment;
import com.aiplantanalyze.model.Post;
import com.aiplantanalyze.model.User;
import com.aiplantanalyze.repository.CommentRepository;
import com.aiplantanalyze.repository.PostRepository;
import com.aiplantanalyze.repository.UserRepository;
import com.aiplantanalyze.security.AuthService;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import jakarta.servlet.http.HttpServletRequest;
import java.util.*;

@RestController
@RequestMapping("/api/forum")
public class ForumController {

    private final PostRepository postRepository;
    private final CommentRepository commentRepository;
    private final UserRepository userRepository;
    private final AuthService authService;

    public ForumController(PostRepository postRepository,
                           CommentRepository commentRepository,
                           UserRepository userRepository,
                           AuthService authService) {
        this.postRepository = postRepository;
        this.commentRepository = commentRepository;
        this.userRepository = userRepository;
        this.authService = authService;
    }

    @GetMapping
    public ResponseEntity<?> getPosts() {
        try {
            List<Post> posts = postRepository.findAll(Sort.by(Sort.Direction.DESC, "createdAt"));
            List<Map<String, Object>> response = new ArrayList<>();
            for (Post post : posts) {
                response.add(buildPostResponse(post));
            }
            return ResponseEntity.ok(response);
        } catch (Exception ex) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Server error", "error", ex.getMessage()));
        }
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getPostById(@PathVariable String id) {
        try {
            Optional<Post> optionalPost = postRepository.findById(id);
            if (optionalPost.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("message", "Post not found"));
            }
            return ResponseEntity.ok(buildPostResponse(optionalPost.get()));
        } catch (Exception ex) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Server error", "error", ex.getMessage()));
        }
    }

    @PostMapping
    public ResponseEntity<?> createPost(HttpServletRequest request,
                                        @RequestBody Map<String, String> body) {
        try {
            User user = authService.getRequiredUser(request);
            String title = body.get("title");
            String content = body.get("content");
            if (title == null || content == null) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                        .body(Map.of("message", "Title and content are required"));
            }
            Post post = new Post();
            post.setTitle(title);
            post.setContent(content);
            post.setAuthorId(user.getId());
            Post savedPost = postRepository.save(post);
            return ResponseEntity.status(HttpStatus.CREATED).body(buildPostResponse(savedPost));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Server error", "error", ex.getMessage()));
        }
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> updatePost(HttpServletRequest request,
                                        @PathVariable String id,
                                        @RequestBody Map<String, String> body) {
        try {
            User user = authService.getRequiredUser(request);
            Optional<Post> optionalPost = postRepository.findById(id);
            if (optionalPost.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("message", "Post not found"));
            }
            Post post = optionalPost.get();
            if (!Objects.equals(post.getAuthorId(), user.getId())) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                        .body(Map.of("message", "Not authorized to update this post"));
            }
            if (body.containsKey("title")) {
                post.setTitle(body.get("title"));
            }
            if (body.containsKey("content")) {
                post.setContent(body.get("content"));
            }
            post.setUpdatedAt(new Date());
            Post updatedPost = postRepository.save(post);
            return ResponseEntity.ok(buildPostResponse(updatedPost));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Server error", "error", ex.getMessage()));
        }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> deletePost(HttpServletRequest request, @PathVariable String id) {
        try {
            User user = authService.getRequiredUser(request);
            Optional<Post> optionalPost = postRepository.findById(id);
            if (optionalPost.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("message", "Post not found"));
            }
            Post post = optionalPost.get();
            if (!Objects.equals(post.getAuthorId(), user.getId())) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                        .body(Map.of("message", "Not authorized to delete this post"));
            }
            commentRepository.deleteAllById(post.getCommentIds());
            postRepository.deleteById(id);
            return ResponseEntity.ok(Map.of("message", "Post deleted successfully"));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Server error", "error", ex.getMessage()));
        }
    }

    @PostMapping("/{id}/comments")
    public ResponseEntity<?> addComment(HttpServletRequest request,
                                        @PathVariable String id,
                                        @RequestBody Map<String, String> body) {
        try {
            User user = authService.getRequiredUser(request);
            String content = body.get("content");
            if (content == null) {
                return ResponseEntity.status(HttpStatus.BAD_REQUEST)
                        .body(Map.of("message", "Comment content is required"));
            }
            Optional<Post> optionalPost = postRepository.findById(id);
            if (optionalPost.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("message", "Post not found"));
            }
            Comment comment = new Comment();
            comment.setContent(content);
            comment.setAuthorId(user.getId());
            comment.setPostId(id);
            Comment savedComment = commentRepository.save(comment);
            Post post = optionalPost.get();
            post.getCommentIds().add(savedComment.getId());
            postRepository.save(post);
            return ResponseEntity.status(HttpStatus.CREATED).body(buildCommentResponse(savedComment));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Server error", "error", ex.getMessage()));
        }
    }

    @DeleteMapping("/{id}/comments/{commentId}")
    public ResponseEntity<?> deleteComment(HttpServletRequest request,
                                           @PathVariable String id,
                                           @PathVariable String commentId) {
        try {
            User user = authService.getRequiredUser(request);
            Optional<Comment> optionalComment = commentRepository.findById(commentId);
            if (optionalComment.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("message", "Comment not found"));
            }
            Comment comment = optionalComment.get();
            if (!Objects.equals(comment.getAuthorId(), user.getId())) {
                return ResponseEntity.status(HttpStatus.FORBIDDEN)
                        .body(Map.of("message", "Not authorized to delete this comment"));
            }
            Optional<Post> optionalPost = postRepository.findById(id);
            if (optionalPost.isPresent()) {
                Post post = optionalPost.get();
                post.getCommentIds().remove(commentId);
                postRepository.save(post);
            }
            commentRepository.deleteById(commentId);
            return ResponseEntity.ok(Map.of("message", "Comment deleted successfully"));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Server error", "error", ex.getMessage()));
        }
    }

    @PutMapping("/{id}/like")
    public ResponseEntity<?> likePost(HttpServletRequest request, @PathVariable String id) {
        try {
            User user = authService.getRequiredUser(request);
            Optional<Post> optionalPost = postRepository.findById(id);
            if (optionalPost.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("message", "Post not found"));
            }
            Post post = optionalPost.get();
            if (post.getLikes().contains(user.getId())) {
                post.getLikes().remove(user.getId());
            } else {
                post.getLikes().add(user.getId());
            }
            postRepository.save(post);
            return ResponseEntity.ok(Map.of("likes", post.getLikes(), "likeCount", post.getLikes().size()));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Server error", "error", ex.getMessage()));
        }
    }

    @PutMapping("/{id}/comments/{commentId}/like")
    public ResponseEntity<?> likeComment(HttpServletRequest request,
                                         @PathVariable String id,
                                         @PathVariable String commentId) {
        try {
            User user = authService.getRequiredUser(request);
            Optional<Comment> optionalComment = commentRepository.findById(commentId);
            if (optionalComment.isEmpty()) {
                return ResponseEntity.status(HttpStatus.NOT_FOUND)
                        .body(Map.of("message", "Comment not found"));
            }
            Comment comment = optionalComment.get();
            if (comment.getLikes().contains(user.getId())) {
                comment.getLikes().remove(user.getId());
            } else {
                comment.getLikes().add(user.getId());
            }
            commentRepository.save(comment);
            return ResponseEntity.ok(Map.of("likes", comment.getLikes(), "likeCount", comment.getLikes().size()));
        } catch (IllegalArgumentException ex) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                    .body(Map.of("message", ex.getMessage()));
        } catch (Exception ex) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("message", "Server error", "error", ex.getMessage()));
        }
    }

    private Map<String, Object> buildPostResponse(Post post) {
        Map<String, Object> response = new HashMap<>();
        response.put("id", post.getId());
        response.put("title", post.getTitle());
        response.put("content", post.getContent());
        response.put("author", buildUserSummary(post.getAuthorId()));
        response.put("likes", post.getLikes());
        response.put("likeCount", post.getLikes().size());

        List<Map<String, Object>> comments = new ArrayList<>();
        for (String commentId : post.getCommentIds()) {
            commentRepository.findById(commentId).ifPresent(comment -> comments.add(buildCommentResponse(comment)));
        }
        response.put("comments", comments);
        response.put("createdAt", post.getCreatedAt());
        response.put("updatedAt", post.getUpdatedAt());
        return response;
    }

    private Map<String, Object> buildCommentResponse(Comment comment) {
        Map<String, Object> response = new HashMap<>();
        response.put("id", comment.getId());
        response.put("content", comment.getContent());
        response.put("author", buildUserSummary(comment.getAuthorId()));
        response.put("likes", comment.getLikes());
        response.put("likeCount", comment.getLikes().size());
        response.put("createdAt", comment.getCreatedAt());
        response.put("updatedAt", comment.getUpdatedAt());
        return response;
    }

    private Map<String, Object> buildUserSummary(String userId) {
        Map<String, Object> author = new HashMap<>();
        if (userId == null) {
            author.put("username", "Unknown");
            return author;
        }
        Optional<User> optionalUser = userRepository.findById(userId);
        author.put("_id", userId);
        author.put("username", optionalUser.map(User::getUsername).orElse("Unknown"));
        return author;
    }
}
